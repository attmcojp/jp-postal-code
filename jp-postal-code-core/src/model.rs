/// 郵便番号レコード
///
/// オリジナルの情報を残すようなモデルにしている。
///
/// # Example
/// ```rust
/// # use jp_postal_code_core::model::UtfKenAllRecord;
/// let _ = UtfKenAllRecord {
///     local_government_code: "01101".to_string(),
///     old_postal_code: "060  ".to_string(),
///     postal_code: "0600000".to_string(),
///     prefecture_kana: "ホッカイドウ".to_string(),
///     city_kana: "サッポロシチュウオウク".to_string(),
///     town_kana: "イカニケイサイガナイバアイ".to_string(),
///     prefecture: "北海道".to_string(),
///     city: "札幌市中央区".to_string(),
///     town: "以下に掲載がない場合".to_string(),
///     has_multi_postal_code: 0,
///     has_chome: 0,
///     has_multi_town: 0,
///     update_code: 0,
///     update_reason: 0,
/// };
/// ```
///
/// # References
/// - [郵便番号データ（1レコード1行、UTF-8形式）の説明](https://www.post.japanpost.jp/zipcode/dl/utf-readme.html)
///
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
pub struct UtfKenAllRecord {
    /// 全国地方公共団体コード（JIS X0401、X0402）
    pub local_government_code: String,
    /// 旧郵便番号（5桁）
    pub old_postal_code: String,
    /// 郵便番号（7桁）
    pub postal_code: String,
    /// 都道府県名（カタカナ）
    pub prefecture_kana: String,
    /// 市区町村名（カタカナ）
    pub city_kana: String,
    /// 町域名（カタカナ）
    pub town_kana: String,
    /// 都道府県名
    pub prefecture: String,
    /// 市区町村名
    pub city: String,
    /// 町域名
    pub town: String,
    /// 一町域が二以上の郵便番号で表される場合の表示
    ///
    /// - 0: 該当せず
    /// - 1: 該当
    pub has_multi_postal_code: i16,
    /// 丁目を有する町域の場合の表示
    ///
    /// - 0: 該当せず
    /// - 1: 該当
    pub has_chome: i16,
    /// 一つの郵便番号で二以上の町域を表す場合の表示
    ///
    /// - 0: 該当せず
    /// - 1: 該当
    pub has_multi_town: i16,
    /// 更新の表示
    ///
    /// - 0: 変更なし
    /// - 1: 変更あり
    /// - 2: 廃止
    pub update_code: i16,
    /// 更新理由
    ///
    /// - 0: 変更なし
    /// - 1: 市政・区政・町政・分区・政令指定都市施行
    /// - 2: 住居表示の実施
    /// - 3: 区画整理
    /// - 4: 郵便区調整等
    /// - 5: 訂正
    /// - 6: 廃止
    pub update_reason: i16,
}
