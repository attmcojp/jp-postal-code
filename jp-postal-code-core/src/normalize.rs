mod zenkaku;

// 正規表現の定義用マクロ
macro_rules! define {
    ($name:ident, $re:expr) => {
        static $name: LazyLock<Regex> = LazyLock::new(|| Regex::new($re).unwrap());
    };
}
#[macro_use]
mod town;

pub use town::normalize_utf_ken_all_record_town;
