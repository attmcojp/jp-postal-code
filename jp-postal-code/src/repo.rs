use jp_postal_code_core::model::UtfKenAllRecord;

pub const DEFAULT_SEARCH_PAGE_SIZE: usize = 10;

#[derive(Debug, Clone)]
pub struct UtfKenAllRepositorySearchRequest<'a> {
    pub postal_code: &'a str,
    pub page_size: Option<usize>,
    pub page_token: Option<&'a str>,
}

#[derive(Debug, Clone)]
pub struct UtfKenAllRepositorySearchResponse {
    pub records: Vec<UtfKenAllRecord>,
    pub next_page_token: Option<String>,
}

/// 郵便番号データベースを扱うリポジトリ
pub trait UtfKenAllRepository: Clone + Send + Sync {
    type Error: std::error::Error + Send + Sync + 'static;

    /// 郵便番号データベースを置き換える
    fn replace(
        &mut self,
        records: &[UtfKenAllRecord],
    ) -> impl std::future::Future<Output = Result<(), Self::Error>> + Send;

    /// 郵便番号データベースから前方一致でレコードを検索する
    fn search(
        &self,
        req: UtfKenAllRepositorySearchRequest<'_>,
    ) -> impl std::future::Future<Output = Result<UtfKenAllRepositorySearchResponse, Self::Error>> + Send;

    /// 郵便番号データベースの総数をカウントする
    fn count(&self) -> impl std::future::Future<Output = Result<usize, Self::Error>> + Send;
}
