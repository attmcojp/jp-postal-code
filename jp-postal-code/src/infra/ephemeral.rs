use crate::repo::{
    UtfKenAllRepository, UtfKenAllRepositorySearchRequest, UtfKenAllRepositorySearchResponse,
    DEFAULT_SEARCH_PAGE_SIZE,
};
use jp_postal_code_core::model::UtfKenAllRecord;
use std::sync::{Arc, Mutex};

#[derive(Debug, Default, Clone)]
pub struct UtfKenAllRepositoryEphemeral {
    records: Arc<Mutex<Vec<UtfKenAllRecord>>>,
}

impl UtfKenAllRepositoryEphemeral {
    pub fn new(records: Vec<UtfKenAllRecord>) -> Self {
        Self {
            records: Arc::new(Mutex::new(records)),
        }
    }

    pub fn into_inner(self) -> Arc<Mutex<Vec<UtfKenAllRecord>>> {
        self.records
    }
}

impl UtfKenAllRepository for UtfKenAllRepositoryEphemeral {
    type Error = std::convert::Infallible;

    #[tracing::instrument(skip(self, records))]
    async fn replace(&mut self, records: &[UtfKenAllRecord]) -> Result<(), Self::Error> {
        self.records
            .lock()
            .unwrap()
            .splice(0.., records.iter().cloned());
        Ok(())
    }

    #[tracing::instrument(skip(self))]
    async fn search(
        &self,
        req: UtfKenAllRepositorySearchRequest<'_>,
    ) -> Result<UtfKenAllRepositorySearchResponse, Self::Error> {
        let page_size = req.page_size.unwrap_or(DEFAULT_SEARCH_PAGE_SIZE);
        let offset = req
            .page_token
            .and_then(|token| token.parse::<usize>().ok())
            .unwrap_or(0);
        let records = self
            .records
            .lock()
            .unwrap()
            .iter()
            .filter(|r| r.postal_code.starts_with(req.postal_code))
            .skip(offset)
            .take(page_size)
            .map(Clone::clone)
            .collect::<Vec<_>>();
        let next_page_token = if records.len() == page_size {
            Some((offset + page_size).to_string())
        } else {
            None
        };
        Ok(UtfKenAllRepositorySearchResponse {
            records,
            next_page_token,
        })
    }

    #[tracing::instrument(skip(self))]
    async fn count(&self) -> Result<usize, Self::Error> {
        Ok(self.records.lock().unwrap().len())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn utf_ken_all_repository_ephemeral_replace() {
        let mut repository = UtfKenAllRepositoryEphemeral::default();

        // 正しくデータが挿入されているかチェック
        repository
            .replace(&[
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "060  ".to_string(),
                    postal_code: "0600000".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "イカニケイサイガナイバアイ".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "以下に掲載がない場合".to_string(),
                    has_multi_postal_code: 0,
                    has_chome: 0,
                    has_multi_town: 0,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "064  ".to_string(),
                    postal_code: "0640941".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "アサヒガオカ".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "旭ケ丘".to_string(),
                    has_multi_postal_code: 0,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "060  ".to_string(),
                    postal_code: "0600041".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "オオドオリヒガシ".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "大通東".to_string(),
                    has_multi_postal_code: 0,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "060  ".to_string(),
                    postal_code: "0600042".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "オオドオリニシ（１−１９チョウメ）".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "大通西（１〜１９丁目）".to_string(),
                    has_multi_postal_code: 1,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "064  ".to_string(),
                    postal_code: "0640820".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "オオドオリニシ（２０−２８チョウメ）".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "大通西（２０〜２８丁目）".to_string(),
                    has_multi_postal_code: 1,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
            ])
            .await
            .unwrap();
        insta::assert_debug_snapshot!(repository.records.lock().unwrap());

        // 正しく置き換えられるかチェック
        repository
            .replace(&[
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "064  ".to_string(),
                    postal_code: "0640941".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "アサヒガオカ".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "旭ケ丘".to_string(),
                    has_multi_postal_code: 0,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "060  ".to_string(),
                    postal_code: "0600042".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "オオドオリニシ（１−１９チョウメ）".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "大通西（１〜１９丁目）".to_string(),
                    has_multi_postal_code: 1,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
            ])
            .await
            .unwrap();
        insta::assert_debug_snapshot!(repository.records.lock().unwrap());
    }

    #[tokio::test]
    async fn utf_ken_all_repository_ephemeral_search() {
        let repository = UtfKenAllRepositoryEphemeral::new(vec![
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600000".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "イカニケイサイガナイバアイ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "以下に掲載がない場合".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 0,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "064  ".to_string(),
                postal_code: "0640941".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "アサヒガオカ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "旭ケ丘".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600041".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリヒガシ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通東".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600042".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリニシ（１−１９チョウメ）".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通西（１〜１９丁目）".to_string(),
                has_multi_postal_code: 1,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "064  ".to_string(),
                postal_code: "0640820".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリニシ（２０−２８チョウメ）".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通西（２０〜２８丁目）".to_string(),
                has_multi_postal_code: 1,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
        ]);

        // 完全一致で検索
        let response = repository
            .search(UtfKenAllRepositorySearchRequest {
                postal_code: "0640820",
                page_size: None,
                page_token: None,
            })
            .await
            .unwrap();
        insta::assert_debug_snapshot!(response);

        // 前方一致で検索
        let response = repository
            .search(UtfKenAllRepositorySearchRequest {
                postal_code: "060",
                page_size: None,
                page_token: None,
            })
            .await
            .unwrap();
        insta::assert_debug_snapshot!(response);
    }

    #[tokio::test]
    async fn utf_ken_all_repository_ephemeral_search_page_size_and_page_token() {
        let repository = UtfKenAllRepositoryEphemeral::new(vec![
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600000".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "イカニケイサイガナイバアイ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "以下に掲載がない場合".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 0,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "064  ".to_string(),
                postal_code: "0640941".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "アサヒガオカ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "旭ケ丘".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600041".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリヒガシ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通東".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600042".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリニシ（１−１９チョウメ）".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通西（１〜１９丁目）".to_string(),
                has_multi_postal_code: 1,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "064  ".to_string(),
                postal_code: "0640820".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリニシ（２０−２８チョウメ）".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通西（２０〜２８丁目）".to_string(),
                has_multi_postal_code: 1,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
        ]);

        // 1, 2 件目を取得
        let response = repository
            .search(UtfKenAllRepositorySearchRequest {
                postal_code: "060",
                page_size: Some(2),
                page_token: None,
            })
            .await
            .unwrap();
        insta::assert_debug_snapshot!(response);

        // 3 件目を取得
        let response = repository
            .search(UtfKenAllRepositorySearchRequest {
                postal_code: "060",
                page_size: Some(2),
                page_token: response.next_page_token.as_deref(),
            })
            .await
            .unwrap();
        insta::assert_debug_snapshot!(response);
    }

    #[tokio::test]
    async fn utf_ken_all_repository_ephemeral_count() {
        let mut repository = UtfKenAllRepositoryEphemeral::new(vec![]);
        let count = repository.count().await.unwrap();
        assert_eq!(count, 0);

        // サンプルデータを入力
        repository
            .replace(&[
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "060  ".to_string(),
                    postal_code: "0600000".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "イカニケイサイガナイバアイ".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "以下に掲載がない場合".to_string(),
                    has_multi_postal_code: 0,
                    has_chome: 0,
                    has_multi_town: 0,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "064  ".to_string(),
                    postal_code: "0640941".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "アサヒガオカ".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "旭ケ丘".to_string(),
                    has_multi_postal_code: 0,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "060  ".to_string(),
                    postal_code: "0600041".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "オオドオリヒガシ".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "大通東".to_string(),
                    has_multi_postal_code: 0,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "060  ".to_string(),
                    postal_code: "0600042".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "オオドオリニシ（１−１９チョウメ）".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "大通西（１〜１９丁目）".to_string(),
                    has_multi_postal_code: 1,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
                UtfKenAllRecord {
                    local_government_code: "01101".to_string(),
                    old_postal_code: "064  ".to_string(),
                    postal_code: "0640820".to_string(),
                    prefecture_kana: "ホッカイドウ".to_string(),
                    city_kana: "サッポロシチュウオウク".to_string(),
                    town_kana: "オオドオリニシ（２０−２８チョウメ）".to_string(),
                    prefecture: "北海道".to_string(),
                    city: "札幌市中央区".to_string(),
                    town: "大通西（２０〜２８丁目）".to_string(),
                    has_multi_postal_code: 1,
                    has_chome: 0,
                    has_multi_town: 1,
                    update_code: 0,
                    update_reason: 0,
                },
            ])
            .await
            .unwrap();

        let count = repository.count().await.unwrap();
        assert_eq!(count, 5);
    }
}
