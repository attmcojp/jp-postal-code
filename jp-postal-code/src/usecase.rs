use crate::repo::UtfKenAllRepository;
use jp_postal_code_core::model::UtfKenAllRecord;
use jp_postal_code_core::normalize::normalize_utf_ken_all_record_town;
use jp_postal_code_util::{download, parse_utf_ken_all_zip, UTF_KEN_ALL_URL};

/// 郵便番号データベースを更新する
#[tracing::instrument(skip(repo, utf_ken_all_zip_url))]
pub async fn update_postal_code_database<R, S>(
    repo: &mut R,
    utf_ken_all_zip_url: Option<S>,
) -> Result<(), anyhow::Error>
where
    R: UtfKenAllRepository,
    S: Into<String>,
{
    let utf_ken_all_zip_url = {
        if let Some(utf_ken_all_zip_url) = utf_ken_all_zip_url {
            utf_ken_all_zip_url.into()
        } else {
            UTF_KEN_ALL_URL.into()
        }
    };
    let mut tempfile = tempfile::tempfile()?;
    tracing::info!(
        ?utf_ken_all_zip_url,
        ?tempfile,
        "Download utf_ken_all.zip into a temp file"
    );
    download(utf_ken_all_zip_url, &mut tempfile).await?;
    tracing::info!(?tempfile, "Parse utf_ken_all.zip to records");
    let records = parse_utf_ken_all_zip(tempfile)?;
    tracing::info!(
        record_count = records.len(),
        "Replace database with the new records"
    );
    repo.replace(&records).await?;
    Ok(())
}

/// 郵便番号を検索する
#[tracing::instrument(skip(repo, postal_code))]
pub async fn search_postal_code<R>(
    repo: &R,
    postal_code: impl AsRef<str>,
) -> Result<Vec<UtfKenAllRecord>, anyhow::Error>
where
    R: UtfKenAllRepository,
{
    let postal_code = postal_code.as_ref();
    let records = repo.search(postal_code).await?;
    let records = records
        .into_iter()
        .flat_map(|r| {
            let towns = normalize_utf_ken_all_record_town(&r);
            towns
                .into_iter()
                .map(|town| UtfKenAllRecord { town, ..r.clone() })
                .collect::<Vec<_>>()
        })
        .collect();
    Ok(records)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::infra::ephemeral::UtfKenAllRepositoryEphemeral;

    #[tokio::test]
    async fn test_update_utf_ken_all_database() {
        let mut server = mockito::Server::new_async().await;
        let mock = server
            .mock("GET", "/zipcode/dl/utf/zip/utf_ken_all.zip")
            .with_status(200)
            .with_header("content-type", "application/zip")
            .with_body_from_file("./testdata/partial_utf_ken_all.zip")
            .create_async()
            .await;

        let mut repo = UtfKenAllRepositoryEphemeral::default();
        update_postal_code_database(
            &mut repo,
            Some(server.url() + "/zipcode/dl/utf/zip/utf_ken_all.zip"),
        )
        .await
        .unwrap();
        mock.assert_async().await;
        insta::assert_debug_snapshot!(repo.into_inner().lock().unwrap());
    }

    #[tokio::test]
    async fn test_search_postal_code() {
        let repo = UtfKenAllRepositoryEphemeral::new(vec![
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600000".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "イカニケイサイガナイバアイ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "以下に掲載がない場合".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 0,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "064  ".to_string(),
                postal_code: "0640941".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "アサヒガオカ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "旭ケ丘".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600041".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリヒガシ".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通東".to_string(),
                has_multi_postal_code: 0,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "060  ".to_string(),
                postal_code: "0600042".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリニシ（１−１９チョウメ）".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通西（１〜１９丁目）".to_string(),
                has_multi_postal_code: 1,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
            UtfKenAllRecord {
                local_government_code: "01101".to_string(),
                old_postal_code: "064  ".to_string(),
                postal_code: "0640820".to_string(),
                prefecture_kana: "ホッカイドウ".to_string(),
                city_kana: "サッポロシチュウオウク".to_string(),
                town_kana: "オオドオリニシ（２０−２８チョウメ）".to_string(),
                prefecture: "北海道".to_string(),
                city: "札幌市中央区".to_string(),
                town: "大通西（２０〜２８丁目）".to_string(),
                has_multi_postal_code: 1,
                has_chome: 0,
                has_multi_town: 1,
                update_code: 0,
                update_reason: 0,
            },
        ]);
        let records = search_postal_code(&repo, "064").await.unwrap();
        insta::assert_debug_snapshot!(records);
    }
}
