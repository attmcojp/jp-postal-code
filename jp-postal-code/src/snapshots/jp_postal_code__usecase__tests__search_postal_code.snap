---
source: jp-postal-code/src/usecase.rs
expression: records
---
[
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640941",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "アサヒガオカ",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "旭ケ丘",
        has_multi_postal_code: 0,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
    UtfKenAllRecord {
        local_government_code: "01101",
        old_postal_code: "064  ",
        postal_code: "0640820",
        prefecture_kana: "ホッカイドウ",
        city_kana: "サッポロシチュウオウク",
        town_kana: "オオドオリニシ（２０−２８チョウメ）",
        prefecture: "北海道",
        city: "札幌市中央区",
        town: "大通西（２０〜２８丁目）",
        has_multi_postal_code: 1,
        has_chome: 0,
        has_multi_town: 1,
        update_code: 0,
        update_reason: 0,
    },
]
